### PR TITLE
paho: use GNRC instead of lwip

### PR DESCRIPTION
**Pull request by OlegHahm**
### Contribution description

The example application for Paho MQTT is using lwip instead of GNRC because at the time of writing no TCP support for GNRC was available. However, in the meantime the situation has changed and with a tiny tweak of the adapter code Paho works like charm on GNRC TCP sockets.

### Testing procedure

Start a MQTT broker, e.g., mosquitto.

On native:
1. Start the example application and add the tab bridge as default route via `nib route add 7 :: <link-local address of the bridge>`

On iotlab-m3 (or similar):
1. Setup a border router (for instance in IoT-Lab following https://www.iot-lab.info/learn/tutorials/riot/riot-public-ipv6-m3/)
2. Start the example application on a second node

Finally, call `con <Broker IPv6 address>` followed by `sub <TOPIC>` and `pub <TOPIC>.

### CODE REVIEW

File: examples/paho-mqtt/Makefile [hunk 1]
```diff
@@ -33,25 +33,20 @@ USEMODULE += ps
...
+USEMODULE += auto_init_gnrc_netif
+# Activate ICMPv6 error messages
+USEMODULE += gnrc_icmpv6_error
+# Specify the mandatory networking module for a IPv6 routing node
+USEMODULE += gnrc_ipv6_router_default
+# Add a routing protocol
+USEMODULE += gnrc_rpl
+USEMODULE += auto_init_gnrc_rpl
+# Additional networking modules that can be dropped if not needed
+USEMODULE += gnrc_icmpv6_echo
```
Comment (benpicco) [inline examples/paho-mqtt/Makefile:45]: Why not support both?
e.g. `examples/gcoap` and [`examples/telnet_server`](https://github.com/RIOT-OS/RIOT/blob/master/examples/telnet_server/Makefile#L14) can be compiled with either GNRC or LWIP.

File: examples/paho-mqtt/Makefile [hunk 2]
```diff
@@ -33,25 +33,20 @@ USEMODULE += ps
...
-  USEMODULE += lwip_ipv6_autoconfig
-endif
-
-USEMODULE += lwip_netdev
-USEMODULE += lwip
+USEMODULE += auto_init_gnrc_netif
+# Activate ICMPv6 error messages
+USEMODULE += gnrc_icmpv6_error
+# Specify the mandatory networking module for a IPv6 routing node
+USEMODULE += gnrc_ipv6_router_default
```
Comment (benpicco) [inline examples/paho-mqtt/Makefile:40]:
```suggestion
USEMODULE += gnrc_ipv6_default
```

should be enough here

File: examples/paho-mqtt/Makefile [hunk 3]
```diff
@@ -33,25 +33,20 @@ USEMODULE += ps
...
+# Add a routing protocol
+USEMODULE += gnrc_rpl
+USEMODULE += auto_init_gnrc_rpl
```
Comment (benpicco) [inline examples/paho-mqtt/Makefile:43]:
```suggestion
```
Not really needed here

File: examples/paho-mqtt/Makefile (hunk 1 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:45]: Sure it would, but what would be the advantage? In my understanding the example applications should be easy to use (and as simple as possible). I think having one example that demonstrates how to build a network application for different stacks is sufficient.

File: examples/paho-mqtt/Makefile (hunk 2 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:40]: I actually thought about which GNRC modules are required and which are optional and came to the conclusion that for the ease-of-use requirement adding routing support is beneficial. Same goes for ICMP. If someone wants to the test the application in the IoT-Lab, for instance, having RPL and routing active might be helpful to actually get a route towards the 6lbr - and having ICMP enabled is beneficial to check the connectivity first.

Of course, we could also remove the modules and try to keep the example application as slim as possible.

Both options have their pros and cons IMO.

File: examples/paho-mqtt/Makefile [hunk 4]
```diff
@@ -33,25 +33,20 @@ USEMODULE += ps
 USEMODULE += netdev_default
 USEPKG += paho-mqtt
 
-# paho-mqtt depends on TCP support, choose which stacks you want
-LWIP_IPV4 ?= 0
```
Comment (aabadie) [inline examples/paho-mqtt/Makefile:37]: Switching to GNRC will drop support for IPv4 from example. I think the use of LwIP stack was not only motivated by the missing TCP support in GNRC but also by the missing IPv4 support.

File: examples/paho-mqtt/Makefile (hunk 2 above)
Comment (benpicco) [reply examples/paho-mqtt/Makefile:40]: The thing is that this will actually make it *harder* to use. When the node is a router, it will not send router solicitations. That means you’ll have to wait for the next unsolicited router advertisement to get a global address.

Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:40]: Valid point. So, how about keeping RPL, but switching to host, i.e., RPL leaf node?

File: examples/paho-mqtt/Makefile (hunk 4 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:37]: To be honest, the main motivation for this PR was that the LWIP version in the IoT-Lab on IPv6 doesn't work for some reason. (The TCP connection is not established.) Since I didn't want to debug LWIP, I tried with GNRC instead.

In the end, the question remains: what is the main use case we want to support here. My idea is that all networking examples use GNRC whenever possible. Of course, that limits the use case to IPv6 (and 6LoWPAN) only. The other option would be to use LWIP for all networking examples wherever possible to allow for both, IPv4 and IPv6.

Comment (benpicco) [reply examples/paho-mqtt/Makefile:37]: But with the sock API we can support both stacks without any changes to the code.
So the Makefile only needs to select different modules, depending on the desired stack.

File: examples/paho-mqtt/Makefile (hunk 2 above)
Comment (benpicco) [reply examples/paho-mqtt/Makefile:40]: tbh I'm not sure if RPL actually is used much at all, at appears to be rather abandoned.

Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:40]: Are you talking about usage or maintenance? As far as I can tell, it is working well.

File: examples/paho-mqtt/Makefile (hunk 4 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:37]: Well, first of all that would require that this application actually works with both stacks which I couldn't verify (LWIP with 6lbr doesn't seem to work). Second, if we go with this reasoning we should update all networking applications which use sock.

File: examples/paho-mqtt/Makefile (hunk 2 above)
Comment (miri64) [reply examples/paho-mqtt/Makefile:40]: If @benpicco isn't using it, it is abandoned ;) I use it regularly in my setups.

Comment (miri64) [reply examples/paho-mqtt/Makefile:40]: (and it is used with gnrc_networking, so most people using that example will use it)

Comment (benpicco) [reply examples/paho-mqtt/Makefile:40]: @miri64 are you only using it with static topologies? It appears that repairing the DODAG  when a node vanishes just doesn't happen, which is kinda crucial in real-world szenarios.
When I asked about it more than a year ago I couldn't find anyone who used it outside simple / short lived test scenarios and nobody who maintains the module.

File: examples/paho-mqtt/Makefile (hunk 4 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:37]: > You can use LWIP without 6lbr, it also supports non-6lo interfaces.

I would prefer an example that works in all scenarios. Having an example application that is known to be broken in a certain configuration may get frustrating for a user.
 
> It's not unreasonable to try to build with both stacks to ensure they work the same. The changes to the `Makefile` are really small.

I'm not per se against this, but what would be the rationale to enable both stacks for this application, but, for instance, not for the MQTT-SN examples?

File: examples/paho-mqtt/Makefile (hunk 2 above)
Comment (OlegHahm) [reply examples/paho-mqtt/Makefile:40]: If this bug is still present basically the same arguments from below (whether to enable LWIP as an alternative or not) apply. If RPL works in some scenarios it is nice to have for this application, but it may get frustrating for users using it an dynamically changing topology. I used RPL on real hardware with bigger topologies only in the IoT-LAB so far, so I've observed the bug myself.

**Comment by OlegHahm** [comment]
Okay, I removed RPL, configure RIOT as a host (instead of router) and let the user select a network stack.

**Review by benpicco — APPROVED** [review]
Looks good to me, please squash

File: examples/paho-mqtt/main.c [hunk 5]
```diff
@@ -293,6 +296,7 @@ static unsigned char readbuf[BUF_SIZE];
 
 int main(void)
 {
+    msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
```
Comment (benpicco) [inline examples/paho-mqtt/main.c:299]: Why is this needed with GNRC?

Comment (miri64) [reply examples/paho-mqtt/main.c:299]: For ping (if that is used), as the executor of the `ping` command is the main thread.

Comment (OlegHahm) [reply examples/paho-mqtt/main.c:299]: My students just learned this the hard way. ;-)

Comment (benpicco) [reply examples/paho-mqtt/main.c:299]: You could put a 

```suggestion
    if (IS_USED(MODULE_GNRC_ICMPV6_ECHO)) {
        msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
    }
```

here then
